core/lib/slider.spec.ts

```diff
@@ -0,0 +1,654 @@
+import type {UnsubscribeFunction, WritableSignal} from '@amadeus-it-group/tansu';
+import {computed, writable} from '@amadeus-it-group/tansu';
+import {afterEach, beforeEach, describe, expect, test, vi} from 'vitest';
+import type {SliderProps, SliderWidget} from './slider';
+import {createSlider} from './slider';
+import type {WidgetState} from './types';
+
+// TODO move to the utils?
+function keyboardEvent(key: string): KeyboardEvent {
+	return {
+		key,
+		preventDefault() {},
+		stopPropagation() {},
+	} as KeyboardEvent;
+}
+
+const defaultStateValues = {
+	min: 0,
+	max: 100,
+	stepSize: 1,
+	values: [0],
+	sortedValues: [0],
+	minValueLabelDisplay: 'hidden',
+	maxValueLabelDisplay: 'visible',
+	combinedLabelDisplay: 'hidden',
+	disabled: false,
+	readonly: false,
+	vertical: false,
+	progressLeft: [0],
+	progressBottom: [0],
+	progressHeight: [100],
+	progressWidth: [0],
+	combinedLabelPositionLeft: 0,
+	combinedLabelPositionTop: 0,
+	handleTooltipLeft: [0],
+	handleTooltipTop: [0],
+	sortedHandles: [{id: 0, value: 0}],
+	className: '',
+	isInteractable: true,
+};
+
+describe(`Slider basic`, () => {
+	let slider: SliderWidget;
+	let state: WidgetState<SliderWidget>;
+	let unsubscribe: UnsubscribeFunction;
+	let defConfig: WritableSignal<Partial<SliderProps>>;
+
+	beforeEach(() => {
+		const sliderElement = document.createElement('div');
+		vi.spyOn(sliderElement, 'getBoundingClientRect').mockReturnValue({x: 10, y: 0, width: 100, height: 4, top: 0, left: 0} as DOMRect);
+		const minLabel = document.createElement('div');
+		vi.spyOn(minLabel, 'getBoundingClientRect').mockReturnValue({x: 10, y: 5, width: 3, height: 4} as DOMRect);
+		const maxLabel = document.createElement('div');
+		vi.spyOn(maxLabel, 'getBoundingClientRect').mockReturnValue({x: 100, y: 5, width: 3, height: 4} as DOMRect);
+
+		defConfig = writable({});
+		slider = createSlider({config: computed(() => ({...defConfig()}))});
+		slider.directives.sliderDirective(sliderElement);
+		slider.directives.minLabelDirective(minLabel);
+		slider.directives.maxLabelDirective(maxLabel);
+		unsubscribe = slider.state$.subscribe((newState) => {
```

Comment:
Note that `beforeEach` can return an `afterEach` function, which would allow you to store the `unsubscribe` variable as a `const` locally in the `beforeEach` function.